File: haskell/fsml/library/Language/Fsml/AST.hs
module Language.Fsml.AST where

data Fsm = Fsm  { states :: [State] }

data State = State { initial     :: Bool
                   , id          :: String
                   , transitions :: [Transition] }
                   deriving (Eq)

data Transition = Transition { input  :: String
                             , action :: Maybe String
                             , target ::  Maybe String }
                             deriving (Eq)

File: scala/fsml/core/src/main/scala/org/softlang/fsml/AST.scala
package org.softlang.fsml

import scala.collection.immutable.Seq

package object AST {

  case class Fsm(states: Seq[State])

  case class State(initial: Boolean, id: String, transitions: Seq[Transition])

  case class Transition(event: String, action: Option[String], target: Option[String])

}
